src/Hydrator.php

```diff
@@ -0,0 +1,245 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Hydrator;
+
+use Closure;
+use Psr\Container\ContainerInterface;
+use ReflectionAttribute;
+use ReflectionClass;
+use ReflectionProperty;
+use Yiisoft\Hydrator\Attribute\SkipHydrate;
+use Yiisoft\Hydrator\Typecast\SimpleTypecast;
+use Yiisoft\Injector\Injector;
+
+use function array_key_exists;
+use function in_array;
+
+/**
+ * @psalm-import-type MapType from HydratorInterface
+ */
+final class Hydrator implements HydratorInterface
+{
+    private Injector $injector;
+    private TypecastInterface $typecast;
+    private DataAttributesHandler $dataAttributesHandler;
+    private ParameterAttributesHandler $parameterAttributesHandler;
+
+    public function __construct(
+        ContainerInterface $container,
+        ?TypecastInterface $typecast = null,
+    ) {
+        $this->injector = new Injector($container);
+        $this->typecast = $typecast ?? new SimpleTypecast();
+        $this->dataAttributesHandler = new DataAttributesHandler($container);
+        $this->parameterAttributesHandler = new ParameterAttributesHandler($container);
+    }
+
+    public function hydrate(object $model, array $data = [], array $map = [], bool $strict = false): void
+    {
+        $this->populate(
+            $model,
+            $this->getHydrateData($model, $data, $map, $strict),
+        );
+    }
+
+    public function create(string $class, array $data = [], array $map = [], bool $strict = false): object
+    {
+        [$excludeProperties, $constructorArguments] = $this->getConstructorArguments($class, $data, $map, $strict);
+
+        $model = $this->injector->make($class, $constructorArguments);
+
+        $this->populate(
+            $model,
+            $this->getHydrateData($model, $data, $map, $strict, $excludeProperties),
+        );
+
+        return $model;
+    }
+
+    /**
+     * @psalm-param class-string $class
+     * @psalm-param MapType $map
+     * @psalm-return array{0:list<string>,1:array<string,mixed>}
+     */
+    private function getConstructorArguments(string $class, array $sourceData, array $map, bool $strict): array
+    {
+        $excludeParameterNames = [];
+        $constructorArguments = [];
+
+        $constructor = (new ReflectionClass($class))->getConstructor();
```

Comment:
How it can be not a string if it is in a method signature `string $class`.

Reply:
> Not handled exception. What if $class is not a class string?

In any case will be exception. May be later we can create friendly exception.

> How it can be not a string if it is in a method signature string $class.
`string`, but no `class-string`

Reply:
#5.